src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,49 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> tail;
+    private int size;
 
+    public void push(T value) {
+        tail = new Node<>(tail, value);
+        size++;
     }
 
     public T peek() {
-        return null;
+        return stackIsEmpty() ? tail.value : null;
     }
 
     public T pop() {
-        return null;
+        T value = null;
+        if (stackIsEmpty()) {
+            value = tail.value;
+            Node<T> oldNode = tail;
+            tail = oldNode.prev;
+        }
+        size--;
+        return value;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T value) {
+            this.prev = prev;
+            this.value = value;
+        }
+    }
+
+    private boolean stackIsEmpty() {
+        if (size != 0) {
+            return true;
+        }
```

Comment:
your method returns true if stack isn't empty , let's rename it (stackIsNotEmpty, checkSize, checkIfNotEmpty)